frontend/src/components/viewers/MetricsVisualizations.tsx

```diff
@@ -344,14 +362,227 @@ type ConfidenceMetric = {
   recall: number;
 };
 
+export interface ConfidenceMetricsFilter {
+  selectedIds: string[];
+  setSelectedIds: (selectedIds: string[]) => void;
+  fullArtifactPathMap: { [key: string]: FullArtifactPath };
+  selectedIdColorMap: { [key: string]: string };
+  setSelectedIdColorMap: (selectedIdColorMap: { [key: string]: string }) => void;
+}
+
 interface ConfidenceMetricsSectionProps {
-  artifacts: Artifact[];
+  linkedArtifacts: LinkedArtifact[];
+  filter?: ConfidenceMetricsFilter;
+}
+
+const runNameCustomRenderer: React.FC<CustomRendererProps<NameId>> = (
+  props: CustomRendererProps<NameId>,
+) => {
+  const runName = props.value ? props.value.name : '';
+  const runId = props.value ? props.value.id : '';
+  return (
+    <Tooltip title={runName} enterDelay={300} placement='top-start'>
+      <Link
+        className={commonCss.link}
+        onClick={e => e.stopPropagation()}
+        to={RoutePage.RUN_DETAILS.replace(':' + RouteParams.runId, runId)}
+      >
+        {runName}
+      </Link>
+    </Tooltip>
+  );
+};
+
+const executionArtifactCustomRenderer: React.FC<CustomRendererProps<string>> = (
+  props: CustomRendererProps<string>,
+) => (
+  <Tooltip title={props.value || ''} enterDelay={300} placement='top-start'>
+    <p className={css.inline}>{props.value || ''}</p>
+  </Tooltip>
+);
+
+const curveLegendCustomRenderer: React.FC<CustomRendererProps<string>> = (
+  props: CustomRendererProps<string>,
+) => (
+  <div
+    style={{
+      width: '2rem',
+      height: '4px',
+      backgroundColor: props.value,
+    }}
+  />
+);
+
+const lineColorsStack = [...lineColors].reverse();
```

Comment:
This is the stack which I use to manage the colors displayed on the ROC Curve plot.

Since there are an unlimited number of artifacts that may be loaded on this page, but only `10` curves can be selected at any given time, I had to create a system that would assign colors to any artifact in a systematic way.

I decided on a stack; this means that a new color is popped off the stack whenever a new artifact is selected, and whenever an artifact is de-selected the associated color is pushed onto the stack. This has a few benefits:
- Since I do not need to generate random colors (which can be [difficult](https://stackoverflow.com/questions/470690/how-to-automatically-generate-n-distinct-colors/4382138#4382138)) for the non-restrained artifact count, I can choose a set number of colors that are distinguishable from one another.
- When a user de-selects and re-selects an artifact, it will keep the same color (and show it on the top of the plot).
- The colors can be moved around in a predictable fashion (provided the user understands the functionality).

This solution was chosen over the alternatives of generating random colors for each artifact and having a set list where artifacts can have the same color (distinguished by hover), as they were too complex or did not present a clear enough UI.

Reply:
Ok awesome!

- Sounds good ✅ 
- Yes, I found this problem when adding tests as well, as the earlier tests were impacted by the later ones. I decided to add the `lineColorsStack` and `setLineColorsStack` as parameters into the `ConfidenceMetricsSection` to fix this issue, as shown in the commit https://github.com/kubeflow/pipelines/pull/8073/commits/d93c0a0809a01ffa409e1c9ca8b71a0c4957128b (which has other changes as well).
Note: I added `lineColorsStack` as a state in `CompareV2` instead of directly to `ConfidenceMetricsSection` because I want the state to be saved after that section is re-rendered when switching between metrics tabs.